jepsen/scalardb/src/scalardb/transfer_append.clj

```diff
@@ -0,0 +1,328 @@
+      (Scan.)
+      (.forNamespace KEYSPACE)
+      (.forTable TABLE)
+      (.withConsistency Consistency/LINEARIZABLE)
+      (.withOrdering (Scan$Ordering. "age" Scan$Ordering$Order/DESC))
+      (.withLimit 1)))
+
+(defn scan-and-return-result
+  [tx scan]
+  (.get (.scan tx scan) 0))
+
+(defn prepare-get-with-age
+  [id age]
+  (-> ((Key. [(IntValue. ACCOUNT_ID id)]) (Key. [(IntValue. AGE age)]))
+      (Get.)
+      (.forNamespace KEYSPACE)
+      (.forTable TABLE)
+      (.withConsistency Consistency/LINEARIZABLE)))
+
+(defn prepare-put
+  [id age balance]
+  (-> (Put. (Key. [(IntValue. ACCOUNT_ID id)]) (Key. [(IntValue. AGE age)]))
+      (.forNamespace KEYSPACE)
+      (.forTable TABLE)
+      (.withValue (IntValue. BALANCE balance))
+      (.withConsistency Consistency/LINEARIZABLE)))
+
+(defn populate-accounts
+  "Insert initial records with transaction.
+  This method assumes that n is small (< 100)"
+  [test n balance]
+  (let [tx (scalar/start-transaction test)]
+    (try
+      (dotimes [i n]
+        (->>
+          (prepare-put i 0 balance)
+          (.put tx)))
+      (.commit tx)
+      (catch Exception e
+        (throw (RuntimeException. (.getMessage e)))))))
+
+(defn calc-new-balance
+  "Calculate the new value from the result of transaction.get()"
+  [^Result result amount]
+  (-> result (.getValue BALANCE) .get .get (+ amount)))
+
+(defn calc-new-age
+  "Calculate the new value from the result of transaction.get()"
+  [^Result result]
+  (-> result (.getValue AGE) .get .get inc))
+
+(defn tx-transfer
+  [tx from to amount]
+  (let [^Result fromResult (scan-and-return-result tx (prepare-scan from))
+        ^Result toResult (scan-and-return-result tx (prepare-scan to))]
+    (->> (prepare-put from (calc-new-age fromResult) (calc-new-balance fromResult (- amount)))
+         (.put tx))
+    (->> (prepare-put to (calc-new-age toResult) (calc-new-balance toResult amount))
+         (.put tx))
+    (.commit tx)))
+
+(defn read-record
+  "Read a record with a transaction. If read fails, this function returns nil."
+  [tx id]
+  (try
+    (first (.scan tx (prepare-scan id)))
+    (catch CrudException _ nil)))
+
+(defn read-all-records
+  "Read records from 0 .. (n - 1)"
+  [test n]
+  (let [tx (scalar/start-transaction test)]
+    (map #(read-record tx %) (range n))))
+
+(defn read-all-with-retry
+  "Read records from 0 .. (n - 1) and retry if needed"
+  [test n]
+  (loop [tries scalar/RETRIES]
+    (when (< tries scalar/RETRIES)
+      (scalar/exponential-backoff (- scalar/RETRIES tries)))
+    (when (zero? (mod tries scalar/RETRIES_FOR_RECONNECTION))
+      (scalar/prepare-transaction-service! test))
+    (let [results (read-all-records test n)]
+      (if (empty? (filter nil? results))
+        results
+        (if (pos? tries)
+          (recur (dec tries))
+          (throw (ex-info "Failed to read all records"
+                          {:cause "Failed to read all records"})))))))
+
+(defn get-balance
```

Comment:
I feel it's less confusing to name `get-balance-from-result`. `get-balance` seems getting balance from db.
The same for the get-version.

Reply:
Good point! Thank you. Renamed.